backend/director/agents/transcription.py

```diff
@@ -0,0 +1,86 @@
+from director.core.session import TextContent, MsgStatus
+from director.tools.videodb_tool import VideoDBTool
+
+logger = logging.getLogger(__name__)
+
+class TranscriptionAgent(BaseAgent):
+    def __init__(self, session=None, **kwargs):
+        self.agent_name = "video_transcription"
+        self.description = (
+            "This is an agent to get transcripts of videos"
+        )
+        self.parameters = self.get_parameters()
+        super().__init__(session=session, **kwargs)
+
+    def run(self, collection_id: str, video_id: str, timestamp_mode: bool = False, time_range: int = 2) -> AgentResponse:
+        """
+        Transcribe a video and optionally format it with timestamps.
+
+        :param str collection_id: The collection_id where given video_id is available.
+        :param str video_id: The id of the video for which the transcription is required.
+        :param bool timestamp_mode: Whether to include timestamps in the transcript.
+        :param int time_range: Time range for grouping transcripts in minutes (default: 2 minutes).
+        :return: AgentResponse with the transcription result.
+        :rtype: AgentResponse
+        """
+        self.output_message.actions.append("Trying to get the video transcription...")
+        output_text_content = TextContent(
+            agent_name=self.agent_name,
+            status_message="Processing the transcription...",
+        )
+        self.output_message.content.append(output_text_content)
+        self.output_message.push_update()
+
+        videodb_tool = VideoDBTool(collection_id=collection_id)
+
+        try:
+            transcript_text = videodb_tool.get_transcript(video_id)
+        except Exception:
+            logger.error("Transcript not found. Indexing spoken words...")
+            self.output_message.actions.append("Indexing spoken words...")
+            self.output_message.push_update()
+            videodb_tool.index_spoken_words(video_id)
+            transcript_text = videodb_tool.get_transcript(video_id)
+
+        if timestamp_mode:
+            self.output_message.actions.append("Formatting transcript with timestamps...")
+            grouped_transcript = self._group_transcript_with_timestamps(
+                transcript_text, time_range
+            )
+            output_text = grouped_transcript
+        else:
+            output_text = transcript_text
+
+        output_text_content.text = output_text
+        output_text_content.status = MsgStatus.success
+        output_text_content.status_message = "Transcription completed successfully."
+        self.output_message.publish()
+
+        return AgentResponse(
+            status=AgentStatus.SUCCESS,
+            message="Transcription successful.",
+            data={"video_id": video_id, "transcript": output_text},
+        )
+    
+    def _group_transcript_with_timestamps(self, transcript_text: str, time_range: int) -> str:
```

Comment:
The grouping logic is not correct.

If you will test it you will get only one block like this:
![image](https://github.com/user-attachments/assets/34151802-d0c6-42b3-9df0-58efbb07a915)

Reason: There are no new lines in transcription text, and even if they were new line representing the given range (2 minutes in case of default) is wrong.

Correct way would be to use the transcription dictionary that VideoDB tool is sending it has timing information unlike the transcription text that is being used.

Reply:
This issue can be resolved by adding free LLM models. Merging [this PR](https://github.com/video-db/Director/pull/80) will fix the problem and it would be helpful for solving similar issues in the future.

Reply:
@sarfarazsiddiquii We have resolved this issue by adding an OpenAI proxy. An OpenAI key is no longer required. Please pull the latest changes from the main branch, ensure that no OpenAI key is present in the .env file, and test the transcript.

Reply:
@ankit-v2-3, Thank you for the update, the code is now testable. 
I’ve fixed the grouping logic in the latest commit, the agent will now properly group the transcription text into 2 minute intervals by default unless time interval is defined.
Let me know if any change is required.
![image](https://github.com/user-attachments/assets/70c75237-c182-4ea7-9530-a6fcd03d20ba)